### PR TITLE
Find semantic errors in front-end instead of BIR verify

### LINKED ISSUE #812: Front-end should not rely on BIR verification
Currently front-end relies on BIR verification to find and report some semantic errors.

The goal of this issue is for the front-end to find and report all semantic errors itself, so that BIR verification can be optional.

Reasons to change our strategy now:
- front-end has richer location information and so can give better error messages
- we are doing unnecessary duplicate work in the verifier
- need for BIR verification is less urgent, since likely to be some time before jBallerina can produce our BIR
- when doing #804, we have to do more semantic checking in codeGen (seeing this already here https://github.com/ballerina-platform/nballerina/commit/259b92297d5f3021de96730862330e1ef56f7272)

### PR DESCRIPTION
**Pull request by ushirask**
\#\# Purpose
> Part of \#812

\#\# Goals
> Describe the solutions that this feature/fix will introduce to resolve the problems described above

\#\# Approach
> Describe how you are implementing the solutions. Include an animated GIF or screenshot if the change affects the UI \(email documentation@wso2.com to review all UI text\). Include a link to a Markdown file or Google doc if the feature write\-up is too long to paste here.

\#\# User stories
> Summary of user stories addressed by this change>

\#\# Release note
> Brief description of the new feature or bug fix as it will appear in the release notes

\#\# Documentation
> Link\(s\) to product documentation that addresses the changes of this PR. If no doc impact, enter “N/A” plus brief explanation of why there’s no doc impact

\#\# Training
> Link to the PR for changes to the training content in https://github.com/wso2/WSO2\-Training, if applicable

\#\# Certification
> Type “Sent” when you have provided new/updated certification questions, plus four answers for each question \(correct answer highlighted in bold\), based on this change. Certification questions/answers should be sent to certification@wso2.com and NOT pasted in this PR. If there is no impact on certification exams, type “N/A” and explain why.

\#\# Marketing
> Link to drafts of marketing content that will describe and promote this feature, including product page changes, technical articles, blog posts, videos, etc., if applicable

\#\# Automation tests
 \- Unit tests 
   > Code coverage information
 \- Integration tests
   > Details about the test cases and coverage

\#\# Security checks
 \- Followed secure coding standards in http://wso2.com/technical\-reports/wso2\-secure\-engineering\-guidelines? yes/no
 \- Ran FindSecurityBugs plugin and verified report? yes/no
 \- Confirmed that this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets? yes/no

\#\# Samples
> Provide high\-level details about the samples related to this feature

\#\# Related PRs
> List any other related PRs

\#\# Migrations \(if applicable\)
> Describe migration steps and platforms on which migration has been tested

\#\# Test environment
> List all JDK versions, operating systems, databases, and browser/versions on which this feature/fix was tested
 
\#\# Learning
> Describe the research phase and any blog posts, patterns, libraries, or add\-ons you used to solve the problem.

### CODE REVIEW

File: compiler/modules/front/codeGen.bal [hunk 1]
```diff
@@ -1230,8 +1230,7 @@ function codeGenConditionalExpr(CodeGenContext cx, bir:BasicBlock block, Environ
 
 function codeGenExprForBoolean(CodeGenContext cx, bir:BasicBlock bb, Environment env, s:Expr expr) returns CodeGenError|BooleanExprEffect {
     var { result, block, narrowing } = check codeGenExpr(cx, bb, env, expr);
-    if result is bir:BooleanOperand {
-        // rest of the type checking is in the verifier
+    if result is bir:BooleanOperand && (result is boolean || t:isSubtypeSimple(result.semType, t:BOOLEAN)) {
```
Comment (manuranga) [inline compiler/modules/front/codeGen.bal]: should be able to remove `result is bir:BooleanOperand &&` since it's redundant.

**Comment by jclark** [comment]
Do this as multiple pull requests.

**Comment by jclark** [comment]
Is there some reason you left this as a draft?

**Comment by jclark** [comment]
Please do function calls next.  Give a more specific location for errors (e.g. specific parameter that is wrong) than we do currently.

**Comment by ushirask** [comment]
> Is there some reason you left this as a draft?

No, should be ready to review.

**Review by jclark — CHANGES_REQUESTED** [review]
Unfortunately I forgot to merge this before I split up codeGen, so this will need updating.

**Review by jclark — CHANGES_REQUESTED** [review]
This looks good but we need to test it.

For each insn in verify.bal that ends up calling operandTypeErr, we should make sure we have a test that checks that the relevant error is caught in the front-end.

I am not sure we have all of them. Consider uses of verifyOperandError for example.

**Review by jclark — CHANGES_REQUESTED** [review]
To test this properly, we really need to have an additional distinct error type that is returned in the "bad BIR" case, and then panic in nback if we get this error type.

File: compiler/modules/comm.err/err.bal [hunk 2]
```diff
@@ -39,3 +41,11 @@ public function illegalArgument(d:Message detail) returns Panic {
 public type SemanticContext object {
     public function semanticErr(d:Message msg, d:Position|d:Range pos, error? cause = ()) returns Semantic;
 };
+
+public function badBIR(d:Message? detail = ()) returns BadBIR {
```
Comment (jclark) [inline compiler/modules/comm.err/err.bal]: Don't add this here. Put in the BIR module as bir:InvalidError

**Review by jclark — CHANGES_REQUESTED** [review]

File: compiler/modules/bir/verify.bal [hunk 3]
```diff
@@ -66,13 +66,24 @@ public function verifyFunctionCode(Module mod, FunctionDefn defn, FunctionCode c
 
 type IntBinaryInsn IntArithmeticBinaryInsn|IntBitwiseBinaryInsn;
 
-function verifyBasicBlock(VerifyContext vc, BasicBlock bb) returns err:Semantic? {
+
+public type InvalidError distinct error;
+
+public function invalidError(d:Message? detail = ()) returns InvalidError {
```
Comment (jclark) [inline compiler/modules/bir/verify.bal]: This should not be public and should be a method on VerifyContext with a Position argument (same as current `err` method)

Rename the existing err method to semanticErr.  Change each `vc.err("bad BIR: XXX"...)` to `vc.invalidErr("XXX"...)`. Change every other vc.err to vc.semanticErr.

**Review by jclark — CHANGES_REQUESTED** [review]

File: compiler/modules/bir/verify.bal [hunk 4]
```diff
@@ -152,147 +158,147 @@ function verifyCall(VerifyContext vc, CallInsn insn) returns err:Semantic? {
...
-        return vc.err("mapping get applied to non-mapping", insn.pos);
+        return vc.semanticErr("mapping get applied to non-mapping", insn.pos);
     }
     t:SemType memberType = t:mappingMemberType(vc.typeContext(), insn.operands[0].semType, k is string ? k : ());
     if insn.name == INSN_MAPPING_GET && (k !is string || !t:mappingMemberRequired(vc.typeContext(), insn.operands[0].semType, k)) {
         memberType = t:union(memberType, t:NIL);
     }
     if !vc.isSameType(memberType, insn.result.semType) {
-        return vc.err(`bad BIR: ${insn.name} result type is not same as member type`, insn.pos);
+        return vc.semanticErr(`bad BIR: ${insn.name} result type is not same as member type`, insn.pos);
```
Comment (jclark) [inline compiler/modules/bir/verify.bal]: Should be invalidErr.  Change this to use `"` quotes.

**Review by jclark — CHANGES_REQUESTED** [review]
Get rid of `bad BIR: ` in the invalidErr messages.

File: compiler/modules/bir/verify.bal [hunk 5]
```diff
@@ -46,18 +46,22 @@ class VerifyContext {
         return self.mod.getPartFile(self.defn.partIndex).qNameRange(startPos);
     }
 
-    function err(d:Message msg, Position|Range pos) returns err:Semantic {
+    function semanticErr(d:Message msg, Position|Range pos) returns err:Semantic {
         return err:semantic(msg, loc=d:location(self.mod.getPartFile(self.defn.partIndex), pos), defnName=self.defn.symbol.identifier);
     }
 
+    function invalidErr(d:Message m, Position|Range pos) returns InvalidError {
+        return error InvalidError("invalid error", (), message=d:messageToString(m), location=d:location(self.mod.getPartFile(self.defn.partIndex), pos), defnName=self.defn.symbol.identifier);
```
Comment (jclark) [inline compiler/modules/bir/verify.bal]: "invalid BIR"

File: compiler/modules/bir/verify.bal [hunk 6]
```diff
@@ -46,18 +46,22 @@ class VerifyContext {
...
+
     function returnType() returns t:SemType => self.defn.signature.returnType;
 
     function symbolToString(Symbol sym) returns string {
         return self.mod.symbolToString(self.defn.partIndex, sym);
     }
 }
 
-public function verifyFunctionCode(Module mod, FunctionDefn defn, FunctionCode code) returns err:Semantic? {
+public function verifyFunctionCode(Module mod, FunctionDefn defn, FunctionCode code) returns err:Semantic|InvalidError? {
```
Comment (jclark) [inline compiler/modules/bir/verify.bal]: Define a union type and use that:

```
type Error err:Semantic|InvalidError;
```

File: compiler/modules/bir/verify.bal [hunk 7]
```diff
@@ -66,13 +70,15 @@ public function verifyFunctionCode(Module mod, FunctionDefn defn, FunctionCode c
 
 type IntBinaryInsn IntArithmeticBinaryInsn|IntBitwiseBinaryInsn;
 
-function verifyBasicBlock(VerifyContext vc, BasicBlock bb) returns err:Semantic? {
+public type InvalidError distinct error<d:SemanticDiagnostic>;
```
Comment (jclark) [inline compiler/modules/bir/verify.bal]: d:SemanticDiagnostic isn't right.

File: compiler/modules/nback/module.bal [hunk 8]
```diff
@@ -53,7 +53,13 @@ public function buildModule(bir:Module birMod, *Options options) returns [llvm:M
     foreach int i in 0 ..< functionDefns.length() {
         bir:FunctionDefn defn = functionDefns[i];
         bir:FunctionCode code = check birMod.generateFunctionCode(i);
-        check bir:verifyFunctionCode(birMod, defn, code);
+        bir:InvalidError|err:Semantic? verifyErr =  bir:verifyFunctionCode(birMod, defn, code);
+        if verifyErr is bir:InvalidError {
+            panic(verifyErr);
```
Comment (jclark) [inline compiler/modules/nback/module.bal]: This won't produce a panic with intelligible locations. We either need to pass it up or turn the locations into something more intelligible here.

**Review by jclark — CHANGES_REQUESTED** [review]

File: compiler/modules/nback/module.bal (hunk 8 above)
Comment (ushirask) [reply compiler/modules/nback/module.bal:58]: Does this mean that we should change the location entry in the InvalidError here before panicking here?

Comment (ushirask) [reply compiler/modules/nback/module.bal]: Does this mean we should change the location entry in the InvalidError before panicking here?

Comment (jclark) [reply compiler/modules/nback/module.bal]: It's means only that the current code is not satisfactory.

Comment (jclark) [reply compiler/modules/nback/module.bal]: We can discuss in the meeting

File: compiler/modules/comm.diagnostic/diagnostic.bal [hunk 9]
```diff
@@ -45,7 +45,8 @@ public function locationLineColumn(Location loc) returns LineColumn {
 public enum Category {
     SYNTAX = "syntax",
     SEMANTIC = "semantic",
-    UNIMPLEMENTED = "unimplemented"
+    UNIMPLEMENTED = "unimplemented",
+    INVALID = "invalid"
```
Comment (jclark) [inline compiler/modules/comm.diagnostic/diagnostic.bal]: internal not invalid

**Review by jclark — CHANGES_REQUESTED** [review]

File: compiler/modules/comm.diagnostic/diagnostic.bal (hunk 9 above)
Comment (ushirask) [reply compiler/modules/comm.diagnostic/diagnostic.bal]: Fixed.

File: compiler/modules/nback/scaffold.bal [hunk 10]
```diff
@@ -4,7 +4,7 @@ import wso2/nballerina.bir;
 import wso2/nballerina.types as t;
 import wso2/nballerina.print.llvm;
 
-type BuildError err:Semantic|err:Unimplemented;
+type BuildError err:Semantic|err:Unimplemented|bir:InvalidError;
```
Comment (jclark) [inline compiler/modules/nback/scaffold.bal]: err:Internal

File: compiler/modules/bir/verify.bal [hunk 11]
```diff
@@ -66,13 +70,17 @@ public function verifyFunctionCode(Module mod, FunctionDefn defn, FunctionCode c
 
 type IntBinaryInsn IntArithmeticBinaryInsn|IntBitwiseBinaryInsn;
 
-function verifyBasicBlock(VerifyContext vc, BasicBlock bb) returns err:Semantic? {
+public type InvalidError distinct error<d:InternalDiagnostic>;
```
Comment (jclark) [inline compiler/modules/bir/verify.bal]: Define this in comm.err as `Internal`.

**Review by jclark — CHANGES_REQUESTED** [review]

**Pull request merged**